File: GoSubmarine/day01/sonarSweep.go
package main

import (
	"fmt"
	"strconv"

	"github.com/tkrehbiel/adventofcode2021/gosubmarine/common"
)

func main() {
	common.Run(common.PuzzleConfig{
		Puzzle:      new(puzzle),
		Day:         1,
		Title:       "Sonar Sweep",
		PuzzleInput: "day01_input.txt",
		TestInput:   []string{"day01_test.txt"},
	})
}

type puzzle struct {
	numbers []int // puzzle input
}

// Common reads the puzzle input into an array of ints
func (p *puzzle) Common(input string) {
	p.numbers = make([]int, 0)
	common.EnumerateLines(input, func(row int, s string) {
		i, err := strconv.Atoi(s)
		if err != nil {
			fmt.Printf("error converting string %s to int: %s\n", s, err)
			panic(err)
		}
		p.numbers = append(p.numbers, i)
	})
	fmt.Printf("read %d numbers\n", len(p.numbers))
}

// Part1 returns the number of increasing values
func (p *puzzle) Part1(input string) (result int) {
	var lastValue int
	for row, value := range p.numbers {
		if row > 0 && value > lastValue {
			result++
		}
		lastValue = value
	}
	return result
}

// Part2 returns the number of increasing values in a 3-row sliding window
func (p *puzzle) Part2(input string) (result int) {
	var lastValue int
	for row := 0; row < len(p.numbers)-2; row++ {
		value := p.numbers[row] + p.numbers[row+1] + p.numbers[row+2]
		if row > 0 && value > lastValue {
			result++
		}
		lastValue = value
	}
	return result
}

File: KotlinSubmarine/src/main/kotlin/day01/SonarSweep.kt
package day01

import Puzzle
import java.io.File

fun main() {
    SonarSweep().run()
}

// https://adventofcode.com/2021/day/1
class SonarSweep : Puzzle(1, "Sonar Sweep") {
    override val label1 = "increases"
    override val label2 = "3-row increases"

    override val puzzleInput = "day01_input.txt"
    override val testInput = arrayOf("day01_test.txt")

    // Part 1 - Just a simple comparison for each line
    override fun part1(input: String): Long {
        var total = 0L
        var row = 0
        var lastValue = 0
        File(input).forEachLine {
            val i = it.toInt()
            if (row > 0 && i > lastValue) total++
            lastValue = i
            row++
        }
        return total
    }

    // Part 2 - Compare 3-row sum for each line
    override fun part2(input: String): Long {
        // Uses a 3-element array to keep track of the sliding window of values
        // So we don't have to read the file into memory beforehand
        var total = 0L
        var row = 0
        var lastValue = 0
        val previous = IntArray(3)
        File(input).forEachLine {
            val i = it.toInt()
            val tally = previous[0] + previous[1] + previous[2]
            if (row > 2 && tally > lastValue) total++
            lastValue = tally
            previous[2] = previous[1]
            previous[1] = previous[0]
            previous[0] = i
            row++
        }
        return total
    }
}

